composer/trainer/trainer.py

```diff
@@ -108,6 +108,112 @@
+             f"{type(pytorch_schedulers[0]).__name__} scheduler. To step the schedulers every batch, adjust the "
+             "scheduler parameters (e.g. step_size, T_max) to be in terms of batches and set "
+             "`step_schedulers_every_batch` to True, or alternatively use a ComposerScheduler. For more information, "
+             f"see https://docs.mosaicml.com/en/v{composer.__version__}/trainer/schedulers.html."))
+
+    else:
+        step_schedulers_every_batch = True
+
+    return step_schedulers_every_batch
+
+
+def _unpack_compute_training_metrics(compute_training_metrics: bool, model: ComposerModel):
+    if compute_training_metrics:
+        warnings.warn(('Computing model evaluation metrics during training doubles the number of forward passes '
+                       'and may lead to a throughput degradation.'))
+        train_metrics = model.metrics(train=True)
+        if isinstance(train_metrics, Metric):
+            # Forcing metrics to be a MetricCollection simplifies logging results
+            train_metrics = MetricCollection([train_metrics])
+
+        return train_metrics
+    else:
+        return None
+
+
+def _compile_schedulers(
+    schedulers: Union[None, ComposerScheduler, PyTorchScheduler, Sequence[Union[ComposerScheduler, PyTorchScheduler]]],
+    state: State,
+    scale_schedule_ratio: float,
+) -> List[PyTorchScheduler]:
+    compiled_schedulers = []
+    for scheduler in ensure_tuple(schedulers):
+        if isinstance(scheduler, PyTorchScheduler):
+            scale_pytorch_scheduler(scheduler, scale_schedule_ratio)
+            compiled_schedulers.append(scheduler)
+        else:  # it's a composer scheduler
+            compiled_schedulers.append(compile_composer_scheduler(scheduler, state, scale_schedule_ratio))
+
+    return compiled_schedulers
+
+
+def _unpack_dataloader(dataloader: Union[DataSpec, Iterable, dict]) -> DataSpec:
+    if isinstance(dataloader, dict):
+        # treat as kwargs for DataSpec
+        dataloader = DataSpec(**dataloader)
+    if not isinstance(dataloader, DataSpec):
+        dataloader = DataSpec(dataloader)
+
+    data_spec = dataloader
+    if isinstance(data_spec.dataloader, torch.utils.data.DataLoader):
+        if data_spec.dataloader._iterator is not None:
+            raise ValueError(
+                textwrap.dedent("""\
+                The `train_dataloader` has an active iterator. This could occur
+                if `persistent_workers=True` and the dataloader has already been iterated,
+                or if the dataloader is mid-epoch. It is required that the training dataloader
+                does not have an active iterator, so CPU dataset augmentations can be
+                correctly inserted.
+
+                To fix, please do not iterate over the dataloader before passing it into
+                the trainer."""))
+    return data_spec
+
+
```

Comment:
maybe this:
```
  if evaluator.should_eval is None:
                if callable(eval_interval):
                    evaluator.should_eval = eval_interval
                else:
                    evaluator.should_eval = evaluate_periodically(eval_interval)
            if evaluator.subset_num_batches is None:
                evaluator.subset_num_batches = subset_num_batches
```

should actually be a function within the evaluator class? e.g. something like:
```
    def update_eval_interval:
```

```

Reply:
Added a setter for the `eval_interval` to the evaluator classes.